<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Practice" data-toc-modified-id="Practice-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Practice</a></span></li><li><span><a href="#Excercise" data-toc-modified-id="Excercise-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Excercise</a></span></li></ul></div>

In [1]:
import pandas as pd

df = pd.read_csv('diabetes.csv')
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

### Practice

In [2]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [4]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [6]:
df.Outcome.value_counts()

0    500
1    268
Name: Outcome, dtype: int64

In [7]:
268/500

0.536

In [8]:
X = df.drop('Outcome', axis=1)
y = df.Outcome

In [9]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled[:3]

array([[ 0.63994726,  0.84832379,  0.14964075,  0.90726993, -0.69289057,
         0.20401277,  0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575,  0.53090156, -0.69289057,
        -0.68442195, -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, -1.28821221, -0.69289057,
        -1.10325546,  0.60439732, -0.10558415]])

In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, stratify=y, random_state=10, test_size=0.2)

In [11]:
len(X_train)

614

In [12]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

scores = cross_val_score(DecisionTreeClassifier(), X, y, cv=5)
scores

array([0.68831169, 0.64935065, 0.7012987 , 0.79084967, 0.73856209])

In [13]:
scores.mean()

0.7136745607333842

In [15]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(
    base_estimator = DecisionTreeClassifier(),
    n_estimators = 100,
    max_samples=0.8,
    oob_score=True,
    random_state=0)
bag_model.fit(X_train, y_train)
bag_model.oob_score_

0.760586319218241

In [16]:
bag_model.score(X_test, y_test)

0.8051948051948052

In [17]:
bag_model = BaggingClassifier(
    base_estimator = DecisionTreeClassifier(),
    n_estimators = 100,
    max_samples=0.8,
    oob_score=True,
    random_state=0)
scores = cross_val_score(bag_model, X, y, cv=5)
scores.mean()

0.7578728461081402

### Excercise 

1. load heart disease into pandas df
2. remove outliers using z score, z score > 3 or z score < -3
3. convert text columns to numbers using label encoding and one hot encoding
4. apply scaling
5. build a classification model using support vector machine. Use standalone model as well as bagging model and check if you see any difference in the performance.
6. now use decision tree classifier. use standalone model as well as bagging and check if you notice any difference in performance
7. comparing performance of svm and decision tree classifier, figure out where it makes most sense to use bagging and why. use internet to figure out in what conditions bagging works the best.

In [18]:
df_heart = pd.read_csv('heart.csv')
df_heart.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [19]:
from scipy import stats

df_heart = df_heart[(np.abs(stats.zscore(df_heart.select_dtypes(include=np.number))) < 3).all(axis=1)]
df_heart.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  899.000000  899.000000   899.000000  899.000000  899.000000   
mean    53.497219  132.027809   198.005562    0.232481  136.917686   
std      9.456073   17.120895   107.157779    0.422649   25.356740   
min     28.000000   80.000000     0.000000    0.000000   63.000000   
25%     47.000000  120.000000   174.500000    0.000000  120.000000   
50%     54.000000  130.000000   222.000000    0.000000  138.000000   
75%     60.000000  140.000000   266.000000    0.000000  156.000000   
max     77.000000  185.000000   518.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  899.000000    899.000000  
mean     0.861513      0.547275  
std      1.007626      0.498037  
min     -2.000000      0.000000  
25%      0.000000      0.000000  
50%      0.500000      1.000000  
75%      1.500000      1.000000  
max      4.000000      1.000000

In [20]:
X = df_heart.loc[:, df_heart.columns != 'HeartDisease']
y = df_heart['HeartDisease']

In [21]:
dummies = pd.get_dummies(X[['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope']])
X = pd.concat([X, dummies], axis=1)
X.drop(['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope'], axis=1,inplace=True)
X.head()

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  Sex_F  Sex_M  \
0   40        140          289          0    172      0.0      0      1   
1   49        160          180          0    156      1.0      1      0   
2   37        130          283          0     98      0.0      0      1   
3   48        138          214          0    108      1.5      1      0   
4   54        150          195          0    122      0.0      0      1   

   ChestPainType_ASY  ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  \
0                  0                  1                  0                 0   
1                  0                  0                  1                 0   
2                  0                  1                  0                 0   
3                  1                  0                  0                 0   
4                  0                  0                  1                 0   

   RestingECG_LVH  RestingECG_Normal  RestingECG_ST  ExerciseAngina_N  \
0               0                  1              0                 1   
1               0                  1              0                 1   
2               0                  0              1                 1   
3               0                  1              0                 0   
4               0                  1              0                 1   

   ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
0                 0              0              0            1  
1                 0              0              1            0  
2                 0              0              0            1  
3                 1              0              1            0  
4                 0              0              0            1

In [29]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

model_params = {
    'svm': {
        'model': SVC(gamma='auto'),
        'params': {
            'svm__C': [1, 10, 20],
            'svm__kernel': ['rbf', 'linear']
        }
    },
    'tree': {
        'model': DecisionTreeClassifier(),
        'params': {
            'tree__criterion': ['gini', 'entropy'],
            'tree__splitter': ['best', 'random'],
            'tree__max_depth': range(1, 20)
        }
    }
}

In [30]:
from sklearn.pipeline import Pipeline
scores = []

for model_name, mp in model_params.items():
    pipe = Pipeline([
        ('scaler', StandardScaler()),
        (model_name, mp['model'])
    ])
    clf = GridSearchCV(estimator = pipe, param_grid = mp['params'], cv=5, return_train_score=False)
    clf.fit(X, y)
    scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })

In [31]:
df = pd.DataFrame(scores, columns=['model', 'best_score', 'best_params'])
df

model  best_score                                        best_params
0   svm    0.827511                {'svm__C': 1, 'svm__kernel': 'rbf'}
1  tree    0.825289  {'tree__criterion': 'gini', 'tree__max_depth':...

In [62]:
# bagged models
model_params = {
    'svm': {
        'model': BaggingClassifier,
        'kwargs': {'base_estimator': SVC(), 'n_estimators': 100, 'max_samples': 0.8, 'oob_score': True,
                   'random_state': 0},
        'params': {
            'svm__base_estimator__gamma': ['auto'],
            'svm__base_estimator__C': [1, 10, 20],
            'svm__base_estimator__kernel': ['rbf', 'linear']
        }
    },
    'tree': {
        'model': BaggingClassifier,
        'kwargs': {'base_estimator': DecisionTreeClassifier(), 'n_estimators': 100, 'max_samples': 0.8, 'oob_score': True,
                   'random_state': 0},
        'params': {
            'tree__base_estimator__criterion': ['gini', 'entropy'],
            'tree__base_estimator__splitter': ['best', 'random'],
            'tree__base_estimator__max_depth': range(1, 20)
        }
    }
}

In [63]:
bag_scores = []

for model_name, mp in model_params.items():
    pipe = Pipeline([
        ('scaler', StandardScaler()),
        (model_name, mp['model'](**mp['kwargs']))
    ])
    clf = GridSearchCV(estimator = pipe, param_grid = mp['params'], cv=5, return_train_score=False)
    clf.fit(X, y)
    bag_scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })


In [64]:
df_bag = pd.DataFrame(bag_scores, columns=['model', 'best_score', 'best_params'])
df_bag

model  best_score                                        best_params
0   svm    0.829733  {'svm__base_estimator__C': 1, 'svm__base_estim...
1  tree    0.833035  {'tree__base_estimator__criterion': 'entropy',...

Before bagging the Support Vector Machine scored higher in performance compared to Decision Trees. However after bagging we see that the ensemble method Decision Tree scored 83.3% compared to SVM 82.9%. This concludes that bagging is a better fit for ensemble methods.